hazelcast/src/main/java/com/hazelcast/map/impl/query/QueryRunner.java

```diff
@@ -92,32 +96,89 @@ public Result runIndexOrPartitionScanQueryOnOwnedPartitions(Query query)
         if (entries != null) {
             // if results have been returned and partition state has not changed, set the partition IDs
             // so that caller is aware of partitions from which results were obtained.
-            return populateTheResult(query, entries, initialPartitions);
+            return populateNonEmptyResult(query, entries, initialPartitions);
+        } else {
+            // else: if fallback to full table scan also failed to return any results due to migrations,
+            // then return empty result set without any partition IDs set (so that it is ignored by callers).
+            return populateEmptyResult(query, initialPartitions);
+        }
+    }
+
+    // MIGRATION UNSAFE QUERYING - MIGRATION STAMPTS ARE NOT VALIDATED, so assumes a run on partition-thread
+    // for a single partition. If the index is global it won't be asked
+    public Result runPartitionIndexOrPartitionScanQueryOnGivenOwnedPartition(Query query, int partitionId) {
```

Comment:
maybe an assertion like this:  `assert nodeEngine.getPartitionService().isPartitionOwner(partitionId)` can be added to get some early notification against potential development hazards